### PR TITLE
[ORCA] Update compute scalar func cost model

### PR DESCRIPTION
**Pull request by dgkimura**
Issue is that following SQL:
```sql
CREATE TABLE t(i) AS SELECT generate_series(1, 10000000) DISTRIBUTED RANDOMLY;
CREATE OR REPLACE FUNCTION add_one (i int) RETURNS int LANGUAGE plpython3u IMMUTABLE AS $$ return i + 1 $$;
EXPLAIN VERBOSE SELECT i, add_one(i) FROM t;
```

Produced following plan:
```
test=# EXPLAIN (COSTS OFF, VERBOSE) SELECT i, add_one(i) FROM t;
                   QUERY PLAN
------------------------------------------------
 Result
   Output: i, add_one(i)
   ->  Gather Motion 3:1  (slice1; segments: 3)
         Output: i
         ->  Seq Scan on public.t
               Output: i
 Optimizer: Pivotal Optimizer (GPORCA)
(7 rows)
```

Here add_one() function is run serially on coordincator rather than in parallel
on the segments because ORCA did not properly account for the cost of scalar
func.

Following plan was a pruned alternative:
```
                QUERY PLAN
------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   Output: i, (add_one(i))
   ->  Seq Scan on public.t
         Output: i, add_one(i)
 Optimizer: Pivotal Optimizer (GPORCA)
(5 rows)
```

However, due to the extra column projected under the gather, ORCA decided
against picking it.

This commit fixes this by adding a cost model param for compute scalar func and
using it to produce better cost comparisons between plans executing scalar func
on coordinator versus executing scalar func on the segments.

### CODE REVIEW

File: contrib/pg_trgm/expected/pg_trgm_optimizer.out [hunk 1]
```diff
@@ -3946,18 +3946,26 @@ SELECT similarity('Szczecin', 'Warsaw');
...
-                     Index Cond: (city % 'Warsaw'::text)
-                     Filter: (city % 'Warsaw'::text)
+               ->  Redistribute Motion 3:3  (slice2; segments: 3)
+                     Hash Key: city, (similarity(city, 'Warsaw'::text)), ('0.3'::real)
+                     ->  GroupAggregate
+                           Group Key: city, (similarity(city, 'Warsaw'::text)), ('0.3'::real)
+                           ->  Sort
+                                 Sort Key: city, (similarity(city, 'Warsaw'::text)), ('0.3'::real)
+                                 ->  Result
+                                       ->  Redistribute Motion 3:3  (slice3; segments: 3)
```
Comment (gpopt) [inline contrib/pg_trgm/expected/pg_trgm_optimizer.out:3963]: What is the kind of motion? Looks like HashRedistribute, but I don't see the Redistribute keys.

Comment (dgkimura) [reply contrib/pg_trgm/expected/pg_trgm_optimizer.out:3963]: Hmm... good point.  Maybe we should fix up explain to display this better.  Based on the physical plan, it turned out to be a random motion:

```
CPhysicalMotionRandom      
+--CPhysicalFilter      
   |--CPhysicalIndexScan   Index Name: ("restaurants_city_idx"), ...]      
   |  +--CScalarCmp (%)      
   |     |--CScalarIdent "city" (0)      
   |     +--CScalarConst (691678577.000)      
   +--CScalarCmp (%)      
      |--CScalarIdent "city" (0)      
      +--CScalarConst (691678577.000)
```

So, I think the plan is trying to evenly redistribute the data to maximize parallel calls to similarity().

Comment (gpopt) [reply contrib/pg_trgm/expected/pg_trgm_optimizer.out:3963]: Thanks, that makes much more sense.

File: src/backend/gporca/data/dxl/minidump/CPruneColumnsTest/AggTopOfMultipleSetRetFuncs.mdp [hunk 2]
```diff
@@ -137,14 +137,14 @@ select count(*) from ( select unnest(arr) i, unnest(arr) j  from (    select str
...
-          <dxl:Cost StartupCost="0" TotalCost="0.000011" Rows="1.000000" Width="8"/>
+          <dxl:Cost StartupCost="0" TotalCost="0.000111" Rows="1.000000" Width="8"/>
         </dxl:Properties>
         <dxl:GroupingColumns/>
         <dxl:ProjList>
           <dxl:ProjElem ColId="4" Alias="count">
             <dxl:AggFunc AggMdid="0.2803.1.0" AggDistinct="false" AggStage="Final" AggKind="n" AggArgTypes="">
               <dxl:ValuesList ParamType="aggargs">
-              <dxl:Ident ColId="5" ColName="ColRef_0005" TypeMdid="0.20.1.0"/>
+                <dxl:Ident ColId="5" ColName="ColRef_0005" TypeMdid="0.20.1.0"/>
```
Comment (gpopt) [inline src/backend/gporca/data/dxl/minidump/CPruneColumnsTest/AggTopOfMultipleSetRetFuncs.mdp:147]: unnecessary change?

File: src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp [hunk 3]
```diff
@@ -143,6 +145,56 @@ CCostModelGPDB::CostScanOutput(CMemoryPool *,  // mp
...
+	CCost costLocal =
+		CCost(num_rebinds * CostTupleProcessing(rows, width, pcp).Get());
+	CCost costChild = CostChildren(mp, exprhdl, pci, pcp);
+
+	CCost costCompute(0);
+
+	if (exprhdl.Pgexpr())
+	{
+		CGroupProxy gp((*exprhdl.Pgexpr())[1]);
+		if (CScalarProjectList::FHasScalarFunc(gp.PgexprFirst()))
```
Comment (gpopt) [inline src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp:180]: Is it possible to get CScalarProjectList child operator from exprhdl.Pop(1)? 
I think it is better to avoid touching memo in the cost model.

In CScalarProjectList , can we also use CExpressionHandle to derive whether it contains scalar function or not? like how `PexprScalarRep` does?

Comment (dgkimura) [reply src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp]: Thanks gpopt!  That makes sense.  I updated as you suggested to derive using CExpressionHandle.

**Review by gpopt — APPROVED** [review]
+1

**Review by chrishajas — APPROVED** [review]
Looks good to me, nice find!

File: src/backend/gporca/libgpdbcost/src/CCostModelParamsGPDB.cpp [hunk 4]
```diff
@@ -183,6 +183,9 @@ const CDouble CCostModelParamsGPDB::DBitmapScanRebindCost(0.06);
 // see CCostModelGPDB::CostHashJoin() for why this is needed
 const CDouble CCostModelParamsGPDB::DPenalizeHJSkewUpperLimit(10.0);
 
+// default scalar func cost
+const CDouble CCostModelParamsGPDB::DScalarFuncCost(1.0e-04);
```
[the remark this answers is not in the document]
Comment (dgkimura) [reply src/backend/gporca/libgpdbcost/src/CCostModelParamsGPDB.cpp:187]: That's a good question.

I had tried to ballpark it based on the other param values.  For example, I figured it's probably more than table scan per tuple cost (DTableScanCostUnitVal = 5.50e-07), but maybe comparable to per tuple cost of building group aggregate hash table (DHashAggInputTupColumnCostUnitVal = 1.20e-04).  Though in actuality, the function cost can be so variable that I'm not sure I can pick a good number.  I think the critical piece is that the cost on coordinator and a segment is comparable.  But, if you have any suggestions or thoughts, I'm definitely open to that.

[the remark this answers is not in the document]
Comment (chrishajas) [reply src/backend/gporca/libgpdbcost/src/CCostModelParamsGPDB.cpp:187]: Got it, yeah I think that's reasonable

**Comment by dgkimura** [comment]
Merged. Thanks for the reviews!
